### PR TITLE
GSK-2902 GSK-2418 Add a method to execute test suite and upload result to Hub programmatically + result cleanup

### PR DESCRIPTION
**Pull request by kevinmessiaen**
### Added an `upload` method to the `TestResult` in order to allow saving them to the Hub.

The method have the following params:
- label: Option[str], defailt `None` the label of the execution. If `None` the execution date will be the label

Example code to execute:

```python
import os

import giskard

os.environ['OPENAI_API_KEY']='sk-...'

client = giskard.GiskardClient(url='http://localhost:9000', key='gsk-...')
suite_id = 1
suite = giskard.Suite.download(client, 'project_key', suite_id)
result = suite.run()
result.upload(client)

``` 

### Cleanup of test result information sent to up

Removed unused inforamtion:
```
actualSlicesSize: List[int] # Unused
referenceSlicesSize: List[int] # Unused
partialUnexpectedIndexList: List[int] # Unused
unexpectedIndexList: List[int] # Unused
missingCount: int # Unused
missingPercent: float # Unused, computable
unexpectedCount: int # Unused
unexpectedPercent: float # Unused, computable
unexpectedPercentTotal: float # Unused
unexpectedPercentNonmissing: float # Unused, computable
```

## Related Issue

- [GSK-2418 (available on Linear)](https://linear.app/giskard/issue/GSK-2418/add-a-method-to-execute-test-suite-and-upload-result-to-hub)
- [GSK-2902 (available on Linear)](https://linear.app/giskard/issue/GSK-2902/cleanup-testresult-information-that-are-sent-to-the-hub)

## Type of Change

- [ ] 📚 Examples / docs / tutorials / dependencies update
- [ ] 🔧 Bug fix (non-breaking change which fixes an issue)
- [ ] 🥂 Improvement (non-breaking change which improves an existing feature)
- [x] 🚀 New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🔐 Security fix

### CODE REVIEW

File: giskard/client/dtos.py [hunk 1]
```diff
@@ -28,6 +35,66 @@ class TestSuiteDTO(ConfiguredBaseModel):
     function_inputs: List[TestInputDTO]
 
 
+class TestSuiteExecutionResult(str, Enum):
+    IN_PROGRESS = ("IN_PROGRESS",)
```
Comment (Hartorn) [inline giskard/client/dtos.py:39]: Why ( "<str>", ), and "ERROR" ? are the parenthesis needed ?

File: giskard/client/dtos.py [hunk 2]
```diff
@@ -28,6 +35,66 @@ class TestSuiteDTO(ConfiguredBaseModel):
...
+
+
+class SaveSuiteTestExecutionDetailsDTO(ConfiguredBaseModel):
+    inputs: Dict[str, List[str]]
+    outputs: List[str]
+    results: List[TestResultStatusEnum]
+    metadata: Dict[str, List[str]]
+
+
+class SaveSuiteTestExecutionDTO(ConfiguredBaseModel):
```
Comment (Hartorn) [inline giskard/client/dtos.py:63]: That's a lot of data. Everything is needed ?

File: giskard/core/suite.py [hunk 3]
```diff
@@ -123,6 +155,68 @@ def _repr_html_(self):
...
+            results=[self._test_result_to_dto(result) for result in self.results],
+            executionDate=self.execution_date.isoformat(),
+            completionDate=self.completion_date.isoformat(),
+        )
+
+    @staticmethod
+    def _test_result_to_dto(result: SuiteResult):
+        datasets = {dataset.id: dataset for dataset in result.params.values() if isinstance(dataset, Dataset)}
+
+        return SaveSuiteTestExecutionDTO(
```
Comment (Hartorn) [inline giskard/core/suite.py:180]: How do we have so many field 😱

File: giskard/client/dtos.py (hunk 1 above)
Comment (kevinmessiaen) [reply giskard/client/dtos.py:39]: My bad I added unnecessary  commas and the pre-commit reformatted like that

File: giskard/client/dtos.py (hunk 2 above)
Comment (kevinmessiaen) [reply giskard/client/dtos.py:63]: Everything is saved into the Hub. 

I agree that we don't use nor need all those information into the Hub. We might probably want to cleanup everything. (Both from Hub and programatic execution)

File: giskard/core/suite.py (hunk 3 above)
Comment (kevinmessiaen) [reply giskard/core/suite.py:180]: I can work work on this on this PR it's relatively safe to clean this up

**Review by Hartorn — CHANGES_REQUESTED** [review]
I did give this PR a try, and had some issues : 

- I had some issues with suite_id not being set (I added a commit for this)
- I cannot debug uploaded test results (issue with uuid badly formatted ?
- when executing test suite from downloaded one, the labels are different (maybe unrelated)

**Comment by Hartorn** [comment]
Executino in hub
image

Execution in notebook uploaded (see with uuid)
image

Finally download + execution in notebook + uploade (see label of tests)
image

The notebook
[wage_classification.zip](https://github.com/Giskard-AI/giskard/files/14434350/wage_classification.zip)

**Comment by kevinmessiaen** [comment]
@Hartorn Thanks for the feedback, I fixed the issue with the `uuid` being wrongly formatted (yes those are also used for debugging so that's why it fails)

**Comment by Hartorn** [comment]
> [ ](/kevinmessiaen)

I did try it again, but still cannot debug

**Comment by kevinmessiaen** [comment]
> @kevinmessiaen I did try it again, but still cannot debug
> 
> ```
> Internal Server Error
> Query did not return a unique result: 14 results were returned
> ```

Yes it was the wrong id send (`suite_id` instead of `suite_test_id`).

I updated the save so that we can keep track of all generated ids properly

**Review by Hartorn — APPROVED** [review]

**Pull request merged**